mlflow/models/signature.py

```diff
@@ -0,0 +1,449 @@
+        del res["columns"]
+
+    with open(os.path.join(path, example_filename), "w") as f:
+        json.dump(res, f, cls=NumpyEncoder)
+    return example_filename
+
+
+def _base64decode(x):
+    return base64.decodebytes(x.encode("ascii"))
+
+
+def dataframe_from_json(path_or_str, schema: Schema = None,
+                        pandas_orient: str = "split") -> pd.DataFrame:
+    """
+    Parse json into pandas.DataFrame. User can pass schema to ensure correct type parsing and to
+    make any necessary conversions (e.g. string -> binary for binary columns).
+
+    :param path_or_str: Path to a json file or a json string.
+    :param schema: Mlflow schema used when parsing the data.
+    :param pandas_orient: pandas data frame convention used to store the data.
+    :return: pandas.DataFrame.
+    """
+    if schema is not None:
+        dtypes = dict(zip(schema.column_names(), schema.column_types()))
+        df = pd.read_json(path_or_str, orient=pandas_orient, dtype=dtypes)[schema.column_names()]
+        binary_cols = [i for i, x in enumerate(schema.column_types()) if x == DataType.binary]
+
+        for i in binary_cols:
+            col = df.columns[i]
+            df[col] = np.array(df[col].map(_base64decode), dtype=np.bytes_)
+            return df
+    else:
+        return pd.read_json(path_or_str, orient=pandas_orient, dtype=False)
+
+
+def _map_numpy_dtype(col: np.ndarray) -> DataType:
+    if not isinstance(col, np.ndarray):
+        raise TypeError("Expected numpy.ndarray, got '{}'.".format(type(col)))
+    if len(col.shape) > 1:
+        raise MlflowException("Expected 1d array, got array with shape {}".format(col.shape))
+    if col.dtype.kind == "b":
+        return DataType.boolean
+    elif col.dtype.kind == "i" or col.dtype.kind == "u":
+        if col.dtype.itemsize < 4 or col.dtype.kind == "i" and col.dtype.itemsize == 4:
+            return DataType.integer
+        elif col.dtype.itemsize < 8 or col.dtype.kind == "i" and col.dtype.itemsize == 8:
+            return DataType.long
+    elif col.dtype.kind == "f":
+        if col.dtype.itemsize <= 4:
+            return DataType.float
+        elif col.dtype.itemsize <= 8:
+            return DataType.double
+    elif col.dtype.kind == "U":
+        return DataType.string
+    elif col.dtype.kind == "S":
+        return DataType.binary
+    elif col.dtype.kind == "O":
+        first_elem = col[0]
+        if isinstance(first_elem, bytearray) or isinstance(first_elem, bytes) and all(
+                [isinstance(x, bytearray) or isinstance(x, bytes) for x in col]):
+            return DataType.binary
+        elif isinstance(first_elem, str) and all([isinstance(x, str) for x in col]):
+            return DataType.string
+        elif isinstance(first_elem, int) and all([isinstance(x, int) for x in col]):
+            return DataType.long
+        elif isinstance(first_elem, float) and all([isinstance(x, float) for x in col]):
```

Comment:
How do nulls interact with these conditonals?

Reply:
Good question, they return false. I will update the logic. Unclear what to do about integers though. Integers do not have a way to represent NaNs or missing values.